src/Bedrock.Framework/Protocols/ConsumableArrayBufferWriter.cs

```diff
@@ -0,0 +1,230 @@
+        /// </summary>
+        public int UnconsumedWrittenCount => _index - _consumedCount;
+
+        /// <summary>
+        /// Returns the total amount of space within the underlying buffer.
+        /// </summary>
+        public int Capacity => _buffer.Length;
+
+        /// <summary>
+        /// Returns the amount of space available that can still be written into without forcing the underlying buffer to grow or shift the data.
+        /// </summary>
+        public int FreeCapacity => _buffer.Length - _index;
+
+        /// <summary>
+        /// Clears the data written to the underlying buffer.
+        /// </summary>
+        /// <remarks>
+        /// You must clear the <see cref="ConsumedArrayBufferWriter{T}"/> before trying to re-use it.
+        /// </remarks>
+        public void Clear()
+        {
+            Debug.Assert(_buffer.Length >= _index);
+            _buffer.AsSpan(0, _index).Clear();
+            _index = 0;
+            _consumedCount = 0;
+        }
+
+        /// <summary>
+        /// Notifies <see cref="IBufferWriter{T}"/> that <paramref name="count"/> amount of data was written to the output <see cref="Span{T}"/>/<see cref="Memory{T}"/>
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <paramref name="count"/> is negative.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown when attempting to advance past the end of the underlying buffer.
+        /// </exception>
+        /// <remarks>
+        /// You must request a new buffer after calling Advance to continue writing more data and cannot write to a previously acquired buffer.
+        /// </remarks>
+        public void Advance(int count)
+        {
+            if (count < 0)
+                throw new ArgumentException(nameof(count));
+
+            if (_index > _buffer.Length - count)
+                throw new InvalidOperationException($"BufferWriter advanced too far. Capacity: {_buffer.Length}");
+
+            _index += count;
+        }
+
+        /// <summary>
+        /// Notifies <see cref="ConsumableArrayBufferWriter{T}"/> that <paramref name="count"/> amount of data was consumed from the output <see cref="Span{T}"/>/<see cref="Memory{T}"/ and can be overwritten>
+        /// </summary>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <paramref name="count"/> is negative.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown when attempting to consume more data than has been written.
+        /// </exception>
+        /// <remarks>
+        /// Do not access data after it has been consumed.
+        /// </remarks>
+        public void Consume(int count)
+        {
+            if (count < 0)
+                throw new ArgumentException(nameof(count));
+
+            var newConsumedCount = _consumedCount + count;
```

Comment:
Overflow potential? Use the logic from here https://github.com/davidfowl/BedrockFramework/pull/53/files#diff-bfc97fa19047ecc32ab4354ebba28f4cR104